src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs

```diff
@@ -207,4 +207,168 @@ public IEnumerable<CompletionResult> CompleteArgument(string commandName, string
             }
         }
     }
+
+
+    /// <summary>
+    ///     Base class for writing custom Argument Completers
+    /// </summary>
+    /// <para>Derived classes should override <see cref="AddCompletionsFor(string,string,IDictionary)"/> </para>
+    public abstract class ArgumentCompleterBase : IArgumentCompleter
+    {
+        private List<CompletionResult> _results;
+
+        /// <summary>
+        ///     The word to complete
+        /// </summary>
+        protected string WordToComplete { get; private set; }
+
+        /// <summary>
+        ///     The Command Ast for the command to complete
+        /// </summary>
+        protected CommandAst CommandAst { get; private set; }
+
+        private List<CompletionResult> Results => _results ?? (_results = new List<CompletionResult>());
+
+        IEnumerable<CompletionResult> IArgumentCompleter.CompleteArgument(string commandName, string parameterName,
+            string wordToComplete, CommandAst commandAst, IDictionary fakeBoundParameters)
+        {
+            WordToComplete = wordToComplete;
+            CommandAst = commandAst;
+            AddCompletionsFor(commandName, parameterName, fakeBoundParameters);
+            return _results;
+        }
+
+        /// <summary>
+        ///     Override in child class to add completions by calling <see ref="CompleteWith"/>
+        /// </summary>
+        /// <param name="commandName">the command to complete parameters for</param>
+        /// <param name="parameterName">the parameter to complete</param>
+        /// <param name="fakeBoundParameters">previously specified command parameters</param>
+        protected abstract void AddCompletionsFor(string commandName, string parameterName, IDictionary fakeBoundParameters);
+
+        /// <summary>
+        ///     Adds a completion result to the result set
+        /// </summary>
+        /// <param name="completionResult">the completion result to add</param>
+        protected void CompleteWith(CompletionResult completionResult)
+            => Results.Add(item: completionResult);
+
+        /// <summary>
+        ///     Adds a completion result to the result set with the specified parameters
+        /// </summary>
+        /// <param name="text">the text to be used as the auto completion result</param>
+        /// <param name="listItemText">the text to be displayed in a list</param>
+        /// <param name="toolTip">the text for the tooltip with details to be displayed about the object</param>
+        /// <param name="resultType">the type of completion result</param>
+        protected void CompleteWith(string text, string listItemText = null, string toolTip = null,
+            CompletionResultType resultType = CompletionResultType.ParameterValue)
```

Comment:
One of my biggest worries about porting posh-git to be a PowerShell completer is that Git's command line conventions don't necessarily align with PowerShell's. For example, most commands should parse anything after `--` as a file path. I'm not familiar enough to know, but it may be that `ParameterValue` isn't the most reasonable default `CompletionResultType` for my use case. (Does anyone have a link where I can read up on the various types? The [docs](https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.completionresulttype) don't tell me much about how to think about native command arguments.)

With that in mind, it might be useful to make the default configurable, e.g.

```csharp

        protected virtual DefaultCompletionResultType => CompletionResultType.ParameterValue;

        protected void CompleteWith(string text, string listItemText = null, string toolTip = null,
            CompletionResultType? resultType = null)
        {
            var quotedText = QuoteIfSpace(text: text);
            var completionResult = new CompletionResult(completionText: quotedText, listItemText: listItemText ?? text,
                resultType: resultType ?? DefaultCompletionResultType, toolTip: toolTip ?? text);
            Results.Add(item: completionResult);
        }
```

Reply:
I like the regex. Very fast parsing. It does have a tendency to take some time to get all of the context back in my head when working with it though - "write once, read never" ;)

Reply:
@dahlbyk You cannot use this base class for the git completer. It is not intended as such, but it would be interesting to take a stab at that to.

The native commands has a different signature for their completers, and I'm thinking about adding a base class which implements 

```csharp
enum NativeCompletionTokenKind {
    None,
    Command,
    Dash,
    DashDash,
    Text
}

class NativeArgumentCompleterBase  {
    protected string CommandName {get;}
    protected string WordToComplete {get;}
    protected string PreviousToken {get;}
    protected NativeCompletionTokenKind PreviousTokenKind {get;}
    protected NativeCompletionTokenKind TokenKind {get;}
    protected CommandAst Ast {get;}
    protected int CursorPosition {get;}
    
    /// TBD
    /// Helpers for 
  
    /// this is used to register the completer with the engine
    static void RegisterCompleter(NativeArgumentCompleterBase completer);
    /// this is the registered public entry point
    public IList<CompletionResult> CompleteInput(string wordToComplete, 
                               CommandAst ast, int cursorPosition) {
       /// Set useful properties from the AST and cursorPosition
      
       Complete();
    }

    protected abstract void Complete();
  
}


class GitCompleter : NativeArgumentCompleterBase , IModuleAssemblyInitializer {
    /// here we have access to an easier abstraction that the CommandAst
    public Complete() {
         ///
    }    

    /// Register the completer.    
    void IModuleAssemblyInitializer.OnImport() =>RegisterCompleter(typeof(GitCompleter );
}
```

From PowerShell, 


One thing that Would be nice in such a case is some built in support for testing.

I regularly write completers where I have a mocked version of a native executable runner, and if the command line matches one of my named commands, I just return the a result from a file stream or a resource.

That way it is easy to experiment with different parsing/completions without even having the command under test installed.